src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+#if DEBUG
+using System.Diagnostics;
+#endif
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal partial class TemporaryStorageServiceFactory
+    {
+        private sealed class ReferenceCountedDisposable<T> : IDisposable
+            where T : IDisposable
+        {
+            private T _instance;
+            private StrongBox<int> _referenceCount;
+
+            /// <summary>
+            /// Initializes a new reference counting wrapper around an <see cref="IDisposable"/> object.
+            /// </summary>
+            /// <remarks>
+            /// <para>The reference count is initialized to 1.</para>
+            /// </remarks>
+            /// <param name="instance">The object owned by this wrapper.</param>
+            public ReferenceCountedDisposable(T instance)
+            {
+                _instance = instance;
+                _referenceCount = new StrongBox<int>(1);
+            }
+
+            private ReferenceCountedDisposable(T instance, StrongBox<int> referenceCount)
+            {
+                _instance = instance;
+
+                // The reference count has already been incremented for this instance
+                _referenceCount = referenceCount;
+            }
+
+            /// <summary>
+            /// Gets the target object.
+            /// </summary>
+            /// <remarks>
+            /// <para>This call is not valid after <see cref="Dispose"/> is called.</para>
+            /// </remarks>
+            public T Target => _instance;
+
+            /// <summary>
+            /// Increments the reference count for the disposable object, and returns a new disposable reference to it.
+            /// </summary>
+            /// <remarks>
+            /// <para>The returned object is an independent reference to the same underlying object. Disposing of the
+            /// returned value multiple times will only cause the reference count to be decreased once.</para>
```

Comment:
> Because it's not too challenging to get right at this scale. 

That's not an acceptable reason.  Complexity has to justify itself over non complex solutions.  "because we can" isn't a reason.   We have countless examples of unnecessary complexity being something that has significantly impacted us in extremely poor ways over Roslyn's lifetime.  There are times when it is appropriate, and data shows the value we get out of that.  But in the absense of that, we should be going for the simplest, correct, solution we can provide.

In this case, it seems like a simple lock, with a couple of data fields (or one, if you want to just merge the boolean + the count into one field), would be sufficient.  It would then be far more trivial to review, comprehend, and maintain.  

There's nothing wrong with using a lock.  It's what we do almost entirely in the IDE code (that, or just using tasks ordering), and it works very well and is consistent with the rest of the code.  Since we don't use these lock-free/wait-free forms, this means this code is automatically much harder to reason about since it's different from our standard patterns and practices, and practically no one on the IDE routinely codes this way.

Please follow the patterns of the rest of the IDE unless we have data that indicates substantive problems with those approaches that warrant these types of solutions.